### PR TITLE
SONARIAC-1550 S7026: Use ADD to retrieve remote resources

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java [hunk 1]
```diff
@@ -40,9 +41,19 @@ public class RetrieveRemoteResourcesCheck implements IacCheck {
 
   private static final String MESSAGE = "Replace this invocation of %s with the ADD instruction.";
 
+  private static final List<String> WGET_AUTH_FLAGS = List.of("--http-user", "--http-password", "--proxy-user", "--proxy-password", "--load-cookies");
   private static final Predicate<String> WGET_DOWNLOAD_FLAG_PREDICATE = startsWithIgnoreQuotes("-O", "--output-document");
   private static final Predicate<String> URL_PREDICATE = startsWithIgnoreQuotes("http");
 
+  // By setting header manually: -H "Authorization ...", --header "Authorization ..." , -H "X-Auth-Token ..." , --header "X-Auth-Token ...".
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java:48]: Not sure what is the purpose of this comment here. The list of flags below doesn't seems to be related to them.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java [hunk 2]
```diff
@@ -61,8 +72,75 @@ public class RetrieveRemoteResourcesCheck implements IacCheck {
...
+
+  private static final CommandDetector WGET_AUTH_HEADERS_SPACE = CommandDetector.builder()
+    .with("--header")
+    .with(startsWithIgnoreQuotes("Authorization", "X-Auth-Token"))
+    .build();
+
   private static final List<CommandDetector> WGET_DETECTORS = List.of(WGET_DOWNLOAD_FLAG_FIRST_DETECTOR, WGET_URL_FIRST_DETECTOR);
-  private static final List<String> WGET_AUTH_FLAGS = List.of("--http-user", "--http-password", "--proxy-user", "--proxy-password", "--load-cookies");
+
+  // curl -o output.txt https://example.com/resource
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java:86]: Good idea to have a short comment before each of those detectors 👍

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java [hunk 3]
```diff
@@ -90,6 +172,32 @@ private static void checkArgument(CheckContext ctx, CommandDetector detector, Li
...
+    });
+  }
+
+  private static boolean containsCurlAuthenticationFlags(List<ArgumentResolution> args) {
+    var containsSimpleAuthFlag = args.stream().anyMatch(arg -> CURL_AUTH_FLAGS.stream().anyMatch(flag -> arg.value().startsWith(flag)));
+    return containsSimpleAuthFlag || containsCurlAuthByHeader(args);
+  }
+
+  private static boolean containsCurlAuthByHeader(List<ArgumentResolution> args) {
+    return !CURL_AUTH_HEADERS.search(args).isEmpty();
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java:201]: We can see duplicated logic with what we have for `get`.
I believe it is fine, as there are not many identical duplicated lines.
Merging the common logic together would probably obfuscate the code with little benefits.

File: iac-extensions/docker/src/test/java/org/sonar/iac/docker/checks/utils/ShortFlagPredicateTest.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    "--abcdX",
+    "-a",
+    "-abc",
+    "-x",
+    "-xyz",
+    "-ABC"
+  })
+  void shouldPredicateBeFalse(String argument) {
+    assertThat(predicate.test(argument)).isFalse();
+  }
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/docker/src/test/java/org/sonar/iac/docker/checks/utils/ShortFlagPredicateTest.java:60]: I would add a test with just `-` if that makes sense.

**Review by rudy-regazzoni-sonarsource — APPROVED** [review]
LGTM! Good work, I'm curious to see what it will raise on next peach run. Please see my few comments.

File: iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java (hunk 3 above)
Comment (mstachniuk) [reply iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/RetrieveRemoteResourcesCheck.java:201]: Yes, it will complicate more the code.

**Pull request merged**